File: 041_Word Pattern.py
"""
Word Pattern

Given a pattern and a string s, check if s follows the same pattern.
"""

def word_pattern(pattern: str, s: str) -> bool:
    words = s.split()
    if len(pattern) != len(words): return False
    p2w, w2p = {}, {}
    for c, w in zip(pattern, words):
        if c in p2w and p2w[c] != w: return False
        if w in w2p and w2p[w] != c: return False
        p2w[c] = w; w2p[w] = c
    return True


if __name__ == '__main__':
    print('Pattern 1:', word_pattern('abba','dog cat cat dog'))
    print('Pattern 2:', word_pattern('abba','dog cat cat fish'))
    print('Pattern 3:', word_pattern('aaaa','dog cat cat dog'))

# Time Complexity: O(n)
# Space Complexity: O(n)

File: 041_Word Pattern.cpp
/*
Word Pattern

Given a pattern and a string s, check if s follows the same pattern.
Example: pattern = "abba", s = "dog cat cat dog" -> true
*/

#include <bits/stdc++.h>
using namespace std;

bool wordPattern(string pattern, string s) {
    istringstream iss(s);
    vector<string> words;
    string w;
    while (iss >> w) words.push_back(w);
    if (pattern.size() != words.size()) return false;
    unordered_map<char,string> p2w;
    unordered_map<string,char> w2p;
    for (int i = 0; i < (int)pattern.size(); ++i) {
        char c = pattern[i];
        const string &word = words[i];
        if (p2w.count(c) && p2w[c] != word) return false;
        if (w2p.count(word) && w2p[word] != c) return false;
        p2w[c] = word; w2p[word] = c;
    }
    return true;
}

int main() {
    cout << "Pattern 1: " << (wordPattern("abba","dog cat cat dog")?"true":"false") << endl;
    cout << "Pattern 2: " << (wordPattern("abba","dog cat cat fish")?"true":"false") << endl;
    cout << "Pattern 3: " << (wordPattern("aaaa","dog cat cat dog")?"true":"false") << endl;
    return 0;
}

// Time Complexity: O(n) where n = number of words
// Space Complexity: O(n)
